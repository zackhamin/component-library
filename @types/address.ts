import { CountryCode } from './countryCode';

export interface Address {
	addressDetails: {
		line1: string;
		line2: string | null;
		city: string;
		county: string | null;
		country: CountryCode;
		postCode: string;
		isDefaultAddress: boolean;
		addressType: string | null;
		addressNickName: string | null;
	};
	addressContactDetails: {
		firstName: string;
		lastName: string;
		mobileNumber: string | null;
	};
	addressMetadata: {
		addressId: string;
		datetimeCreated: Date | string;
		datetimeUpdated: Date | string;
		addressType: string | null;
	};
}
