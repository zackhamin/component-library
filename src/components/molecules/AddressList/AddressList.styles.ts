import styled from 'styled-components';

const StyledList = styled.ul`
	list-style: none;
	display: flex;
	flex-direction: column;
	gap: 1.6rem;
	padding: 0;
	margin: 0;
`;

export { StyledList };
