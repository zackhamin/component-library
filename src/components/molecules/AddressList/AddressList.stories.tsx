import { Address } from '@Types/address';
import { mockAddress } from 'MockData/address';
import { AddressList, AddressListProps } from './AddressList';

export default {
	title: 'Molecules/AddressList',
	component: AddressList,
};

export const AddressListStory = (args: AddressListProps) => (
	<AddressList {...args} />
);

const addresses: Address[] = [
	{
		...mockAddress,
		addressDetails: {
			...mockAddress.addressDetails,
			line1: '123 Main St',
		},
		addressMetadata: {
			...mockAddress.addressMetadata,
			addressId: 'f8d4e485-3b52-4e79-8ab6-fbdf88b6bd96',
		},
	},
	mockAddress,
	{
		...mockAddress,
		addressContactDetails: {
			...mockAddress.addressContactDetails,
			firstName: 'Jonathan',
			lastName: 'Doesfield',
		},
		addressDetails: {
			...mockAddress.addressDetails,
			line1: '123 Other St',
			line2: 'Top Floor Flat',
			city: 'Llanfairpwllgwyngyll',
			postCode: 'LL61 1AA',
			county: 'Gwynedd',
			country: 'GB',
		},
		addressMetadata: {
			...mockAddress.addressMetadata,
			addressId: 'f8d4e485-3b52-4e79-8ab6-fbdf88b6bd96',
		},
	},
];

AddressListStory.storyName = 'AddressList';
AddressListStory.args = {
	addresses,
	selectedAddress: mockAddress,
};
AddressListStory.argTypes = {
	handleSelect: { action: 'selected' },
};
