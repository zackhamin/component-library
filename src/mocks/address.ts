import { Address } from '@Types/address';

export const mockAddress: Address = {
	addressDetails: {
		line1: '22 some street',
		line2: null,
		city: 'Never land',
		county: null,
		country: 'GB',
		postCode: 'DS2 3CD',
		isDefaultAddress: false,
		addressType: null,
		addressNickName: null,
	},
	addressContactDetails: {
		firstName: 'Zack',
		lastName: 'Amin',
		mobileNumber: '0789892323',
	},
	addressMetadata: {
		addressId: '8d063d44-1e27-4a38-9602-552730b90552',
		datetimeCreated: '2022-01-01 00:00:00 UTC',
		datetimeUpdated: '2022-01-01 00:00:00 UTC',
		addressType: null,
	},
};
