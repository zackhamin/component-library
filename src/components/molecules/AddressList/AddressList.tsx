import { Address as AddressType } from '@Types/address';

import { Address } from 'Atoms/Address';
import { ListOption } from 'Atoms/ListOption';
import { Text } from 'Atoms/Text';

import { StyledList } from './AddressList.styles';

export interface AddressListProps {
	addresses: AddressType[];
	selectedAddress?: AddressType;
	handleSelect: (option: AddressType) => void;
	i18n: {
		strings: {
			noAddresses: string;
		};
	};
}

export const AddressList = ({
	handleSelect,
	addresses,
	selectedAddress,
	i18n = { strings: { noAddresses: '' } },
}: AddressListProps) => {
	return (
		<StyledList>
			{addresses.length ? (
				addresses.map((address) => {
					const isSelected =
						selectedAddress?.addressMetadata.addressId ===
						address.addressMetadata.addressId;

					return (
						<ListOption
							key={address.addressMetadata.addressId}
							isSelected={isSelected}
							handleSelect={() => handleSelect(address)}
						>
							<Address address={address} options={{ showName: true }} />
						</ListOption>
					);
				})
			) : (
				<Text>{i18n.strings.noAddresses}</Text>
			)}
		</StyledList>
	);
};
