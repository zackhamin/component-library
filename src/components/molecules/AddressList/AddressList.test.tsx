import { Address } from '@Types/address';
import { fireEvent, render } from '@testing-library/react';

import { AddressList } from './AddressList';
import { mockAddress } from 'MockData/address';

jest.spyOn(console, 'error').mockImplementation(jest.fn);

const addresses: Address[] = [
	mockAddress,
	{
		...mockAddress,
		addressDetails: {
			...mockAddress.addressDetails,
			line1: '12 High Road',
			line2: 'Flat 1',
		},
		addressMetadata: {
			...mockAddress.addressMetadata,
			addressId: '485ce2f8-ccf3-4719-99a0-4f47ce63253a',
		},
	},
];

export const mocki18n = { strings: { noAddresses: '' } };

describe('AddressList', () => {
	const mockSelectHandler = jest.fn();

	afterEach(() => {
		jest.resetAllMocks();
	});

	it('renders card option correctly', () => {
		const { container } = render(
			<AddressList
				handleSelect={mockSelectHandler}
				addresses={addresses}
				selectedAddress={addresses[0]}
				i18n={mocki18n}
			/>
		);

		expect(container).toMatchSnapshot();
	});

	it('ensures selected address has an active state', () => {
		const { container } = render(
			<AddressList
				handleSelect={mockSelectHandler}
				addresses={addresses}
				selectedAddress={addresses[1]}
				i18n={mocki18n}
			/>
		);

		const option = container.getElementsByTagName('li')[1];

		expect(option).toHaveClass('active');
	});

	it('updates selected address option', () => {
		const { container } = render(
			<AddressList
				handleSelect={mockSelectHandler}
				addresses={addresses}
				i18n={mocki18n}
			/>
		);

		const optionToSelect = container.getElementsByTagName('li')[1];
		expect(optionToSelect).not.toHaveClass('active');
		fireEvent.click(optionToSelect);
		setTimeout(() => {
			expect(optionToSelect).toHaveClass('active');
		});
	});

	it('calls the select handler correctly', () => {
		const { container } = render(
			<AddressList
				handleSelect={mockSelectHandler}
				addresses={addresses}
				selectedAddress={addresses[0]}
				i18n={mocki18n}
			/>
		);

		const methodToSelect = container.getElementsByTagName('li')[1];

		fireEvent.click(methodToSelect);

		expect(mockSelectHandler).toHaveBeenCalled();
	});
});
