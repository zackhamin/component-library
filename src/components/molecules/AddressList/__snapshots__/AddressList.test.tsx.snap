// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressList renders card option correctly 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  list-style: none;
  border: 1px solid #EEEEEF;
  background: #ffffff;
  border-radius: 8px;
  padding: 1.6rem;
  cursor: pointer;
}

.c1.active {
  border: 1px solid #7FC1BA;
  background: rgba(3,134,120,0.1);
}

.c1.active .radio {
  border: 1px solid #00594C;
}

.c1.active .radio > span {
  background: #00594C;
}

.c1 .details {
  margin-left: 2.6rem;
}

.c1 .radio {
  display: block;
  border: 1px solid #bebebe;
  border-radius: 50%;
}

.c1 .radio > span {
  display: block;
  width: 10px;
  height: 10px;
  margin: 3px;
  border-radius: 50%;
  background: transparent;
}

.c0 {
  list-style: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 1.6rem;
  padding: 0;
  margin: 0;
}

<div>
  <ul
    class="c0"
  >
    <li
      class="c1 active"
    >
      <div
        class="radio"
      >
        <span />
      </div>
      <div
        class="details"
      >
        <p>
          Zack
           
          Amin
          ,
           
          22 some street
          , 
          Never land
          , 
          DS2 3CD
          , 
          GB
          <br />
        </p>
      </div>
    </li>
    <li
      class="c1"
    >
      <div
        class="radio"
      >
        <span />
      </div>
      <div
        class="details"
      >
        <p>
          Zack
           
          Amin
          ,
           
          12 High Road
          , 
          Flat 1
          , 
          Never land
          , 
          DS2 3CD
          , 
          GB
          <br />
        </p>
      </div>
    </li>
  </ul>
</div>
`;
